docsTech/agnostic-styling.md

```diff
@@ -0,0 +1,133 @@
+
+### Conditional styles
+
+We currently use to a small extent the `({ ownerState }) => /* dynamic styles */` pattern to apply conditional styles. This pattern is not applicable to CSS stylesheets with the proposed solution because we rely on having static styles. Supporting that pattern would amount to re-implementing PigmentCSS here.
+
+#### Current situation
+
+```javascript.jsx
+const useUtilityClasses = (ownerState) => { /* ... */ }
+
+const GridQuickFilterTrigger = styled(ToolbarButton, {
+  name: 'MuiDataGrid',
+  slot: 'toolbarQuickFilterTrigger',
+})(({ ownerState }: { ownerState: OwnerState }) => ({
+  zIndex: 1,
+  opacity: ownerState.expanded ? 0 : 1,
+  pointerEvents: ownerState.expanded ? 'none' : 'auto',
+  transition: vars.transition(['opacity']),
+}));
+
+// usage
+const classes = useUtilityClasses(rootProps)
+<GridQuickFilterTrigger
+  ownerState={{ ...rootProps, expanded: true }}
+  className={classes.trigger}
+/>
+```
+
+#### Proposed syntax
+
+```javascript.jsx
+const triggerSlot = slot({
+  name: 'MuiDataGrid',
+  slot: 'toolbarQuickFilterTrigger',
+  as: ToolbarButton,
+}, {
+  root: {
+    zIndex: 1,
+    opacity: 1,
+    pointerEvents: 'auto',
+    transition: vars.transition(['opacity']),
+  },
+  expanded: {
+    opacity: 0,
+    pointerEvents: 'none',
+  },
+}));
+
+// usage
+const trigger = useGridStyleSlot(rootProps, triggerSlot)
+<trigger.component
+  ownerState={rootProps}
+  className={clsx(
+    trigger.classes.root,
+    expanded && trigger.classes.expanded,
+  )}
+/>
+```
+
+#### Proposed compilation (emotion)
+
+```javascript.jsx
+const triggerSlot = {
+  component: styled(ToolbarButton, {
+    name: 'MuiDataGrid',
+    slot: 'toolbarQuickFilterTrigger',
+  })({
+    /* root: */
+    zIndex: 1,
+    opacity: 1,
+    pointerEvents: 'auto',
+    transition: vars.transition(['opacity']),
+
+    /* expanded: */
+    '&.MuiDataGrid-toolbarQuickFilterTrigger--expanded': {
+      opacity: 0,
+      pointerEvents: 'none',
+    },
+  }),
+  classes: {
+    root: 'MuiDataGrid-toolbarQuickFilterTrigger',
+    expanded: 'MuiDataGrid-toolbarQuickFilterTrigger--expanded',
+  },
+}
+
+// usage
+const trigger = useGridStyleSlot(rootProps, triggerSlot)
+<trigger.component
+  ownerState={rootProps}
+  className={clsx(
+    trigger.classes.root,
+    expanded && trigger.classes.expanded,
+  )}
+/>
+```
+
+#### Proposed compilation (no emotion)
+
+```javascript.jsx
+import './[currentFilename].autogenerated.css';
+
+const triggerSlot = slot({
+  name: 'MuiDataGrid',
+  slot: 'toolbarQuickFilterTrigger',
+  as: ToolbarButton,
+}, {
+  root: 'MuiDataGrid-toolbarQuickFilterTrigger',
+  expanded: 'MuiDataGrid-toolbarQuickFilterTrigger--expanded',
+}))
+
+// usage
+const trigger = useGridStyleSlot(rootProps, triggerSlot)
+<trigger.component
+  ownerState={rootProps}
+  className={clsx(
+    trigger.classes.root,
+    expanded && trigger.classes.expanded,
+  )}
+/>
+```
+
+#### Downsides
```

Comment:
One more: because we need to keep supporting emotion, we also need to reproduce all of its shitty decisions in our new solution. Emotion doesn't support CSS keyframes in `styled()` CSS JS styles ([example](https://github.com/mui/mui-x/blob/2733ad31cb84913bf6a466929585aab803eac39e/packages/x-data-grid/src/components/GridShadowScrollArea.tsx#L13-L18)), it only supports them in CSS string templates, so we also need to support that syntax in our babel plugin.

Reply:
And we also can't adopt new APIs that `@mui/system` doesn't support like `const className = css({ ... })` -_-